pkg/vfs/fill.go

```diff
@@ -34,36 +34,88 @@ type _file struct {
 	size uint64
 }
 
-func (v *VFS) fillCache(ctx meta.Context, paths []string, concurrent int, count, bytes *uint64) {
-	logger.Infof("start to warmup %d paths with %d workers", len(paths), concurrent)
+type CacheAction uint8
+
+func (act CacheAction) String() string {
+	switch act {
+	case WarmupCache:
+		return "warmup cache"
+	case EvictCache:
+		return "evict cache"
+	case CheckCache:
+		return "check cache"
+	}
+	return "unknown operation"
+}
+
+const (
+	WarmupCache CacheAction = iota
+	EvictCache
+	CheckCache = 2
+)
+
+func (v *VFS) cache(ctx meta.Context, action CacheAction, paths []string, concurrent int, resp *CacheResponse) {
+	logger.Infof("start to %s %d paths with %d workers", action, len(paths), concurrent)
+
 	start := time.Now()
-	todo := make(chan _file, 10240)
+	todo := make(chan _file, 10*concurrent)
 	wg := sync.WaitGroup{}
 	for i := 0; i < concurrent; i++ {
 		wg.Add(1)
 		go func() {
-			for {
-				f := <-todo
-				if f.ino == 0 {
-					break
+			defer wg.Done()
+			for f := range todo {
+				if ctx.Canceled() {
+					return
 				}
-				if err := v.fillInode(ctx, f.ino, f.size, bytes); err != nil {
-					logger.Errorf("Inode %d could be corrupted: %s", f.ino, err)
+
+				if f.ino == 0 {
+					continue
 				}
-				if v.Conf.Meta.OpenCache > 0 {
-					if err := v.Meta.Open(ctx, f.ino, syscall.O_RDONLY, &meta.Attr{}); err != 0 {
-						logger.Errorf("Inode %d could be opened: %s", f.ino, err)
+
+				iter := newSliceIterator(ctx, v.Meta, f.ino, f.size)
+				var handler sliceHandler
+				switch action {
+				case WarmupCache:
+					handler = func(s meta.Slice) error {
+						return v.Store.FillCache(s.Id, s.Size)
+					}
+
+					if v.Conf.Meta.OpenCache > 0 {
+						if err := v.Meta.Open(ctx, f.ino, syscall.O_RDONLY, &meta.Attr{}); err != 0 {
+							logger.Errorf("Inode %d could be opened: %s", f.ino, err)
+						}
+						_ = v.Meta.Close(ctx, f.ino)
+					}
+				case EvictCache:
+					handler = func(s meta.Slice) error {
+						return v.Store.EvictCache(s.Id, s.Size)
+					}
+				case CheckCache:
+					handler = func(s meta.Slice) error {
+						missBytes, err := v.Store.CheckCache(s.Id, s.Size)
+						if err != nil {
```

Comment:
for now err can be ignored.

Reply:
This error handling is kept here, based on the check API.